doctest: initial submission 2.4.6

dankamongmen opened the pull request:
#### Description

onqtam/doctest, needed by Notcurses (#11838)

###### Type(s)

- [X] new submission
- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'printf "%s\n" "macOS `sw_vers -productVersion` `sw_vers -buildVersion` `uname -m`" "`xcodebuild -version|awk '\''NR==1{x=$0}END{print x" "$NF}'\''`"'|tee /dev/tty|pbcopy
-->
macOS 11.4 (XNU 20.5.0)
Xcode x.y

###### Verification <!-- (delete not applicable items) -->
Have you

- [ ] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [X] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [X] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [X] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [X] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [X] tried a full install with `sudo port -vst install`?
- [X] tested basic functionality of all binary files?

reneeotten commented:
@dankamongmen thanks for your submission! 

Unfortunately, there are a few issues with it: one of them that it doesn't build ;) I notice that you skipped a few points in our PR checklist, that's not really recommended. You should certainly build it locally yourself and make sure that this works before submitting a PR. 

Additionally, the formatting of the Portfile doesn't follow our typical recommendations. Please add the default modeline, and conform your whitespacing to that (i.e., multiples of four spaces). Once you have made the necessary adjustments please push again to this PR and we can take a closer look!

reneeotten commented on devel/doctest/Portfile:
```diff
@@ -0,0 +1,18 @@
+PortSystem 1.0
+PortGroup cmake 1.0
+PortGroup github 1.0
+github.setup        onqtam doctest 2.4.6 v
```
there should be no `v` here at the end, since upstream doesn't have this as part of the filenam

dankamongmen replied:
yep, corrected, verified that we now successfully fetch (also needed move from `releases` to bare tags =\, but it works).

reneeotten commented on devel/doctest/Portfile:
```diff
@@ -0,0 +1,18 @@
+PortSystem 1.0
```
please add the usual modeline, which can be copied from another Portfile or the guide

dankamongmen replied:
applied `# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4`

reneeotten commented on devel/doctest/Portfile:
```diff
@@ -0,0 +1,18 @@
+PortSystem 1.0
+PortGroup cmake 1.0
```
please use the latest cmake PG, which is version `1.1`

dankamongmen replied:
oh indeed, thanks, done

reneeotten commented on devel/doctest/Portfile:
```diff
@@ -0,0 +1,18 @@
+long_description \
+ doctest is a light and feature-rich C++98 / C++11 single-header testing \
+ framework for unit tests and TDD.
+homepage https://github.com/onqtam/doctest
```
homepage is not necessary here as it is set automatically by the `github` PG

dankamongmen replied:
yanked, thanks

reneeotten commented on devel/doctest/Portfile:
```diff
@@ -0,0 +1,18 @@
+platforms darwin
+license MIT
+test.run yes
+test.target check
```
you should have the `checksums` listed somewhere here

dankamongmen replied:
of course, done

dankamongmen commented:
oh absolutely! sorry this wasn't meant to be a final submission; I was just building up to one. my apologies for any wasted time =[.

reneeotten commented:
> oh absolutely! sorry this wasn't meant to be a final submission; I was just building up to one. my apologies for any wasted time =[.

okay, thanks for explaining; but then you might misunderstand the purpose of a PR - or at least the way we use it ;) Once you open a PR that signals that it's ready to review, in rare cases you could set it to `draft` but personally I am not a fan of this.... You're supposed to do the legwork locally and once you think everything works as intended you can open a PR.

I will set this to a draft for now, once you've figured out things locally you can push to the PR and mark it as "ready for review".

dankamongmen commented:
alright, sorry again for the annoyance and delay. i've done the minimal fixups to get things properly fetching and building. i'm going through your detailed feedback now. thanks again!

dankamongmen commented:
`port lint` is clear, but `port test` is failing -- i think i need to fix up my `test.target` there.

dankamongmen commented:
yeah, just yanking the `test.target` line completely resolved that, though i'm seeing a few unit tests fail:

```
:info:test Executing:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_devel_doctest/doctest/work/doctest-2.4.6" && /usr/bin/make test 
:debug:test system:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_devel_doctest/doctest/work/doctest-2.4.6" && /usr/bin/make test 
:info:test Running tests...
:info:test /opt/local/bin/ctest --force-new-ctest-process 
:info:test Test project /opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_devel_doctest/doctest/work/doctest-2.4.6
:info:test         Start   1: main.cpp
:info:test   1/135 Test   #1: main.cpp ...................................   Passed    0.08 sec
:info:test         Start   2: main.cpp_xml
:info:test   2/135 Test   #2: main.cpp_xml ...............................   Passed    0.02 sec
:info:test         Start   3: main.cpp_junit
:info:test   3/135 Test   #3: main.cpp_junit .............................   Passed    0.02 sec
:info:test         Start   4: doctest_proxy.h
:info:test   4/135 Test   #4: doctest_proxy.h ............................   Passed    0.02 sec
:info:test         Start   5: doctest_proxy.h_xml
:info:test   5/135 Test   #5: doctest_proxy.h_xml ........................   Passed    0.02 sec
:info:test         Start   6: doctest_proxy.h_junit
:info:test   6/135 Test   #6: doctest_proxy.h_junit ......................   Passed    0.02 sec
:info:test         Start   7: header.h
:info:test   7/135 Test   #7: header.h ...................................   Passed    0.02 sec
:info:test         Start   8: header.h_xml
:info:test   8/135 Test   #8: header.h_xml ...............................   Passed    0.02 sec
:info:test         Start   9: header.h_junit
:info:test   9/135 Test   #9: header.h_junit .............................   Passed    0.02 sec
:info:test         Start  10: alternative_macros.cpp
:info:test  10/135 Test  #10: alternative_macros.cpp .....................   Passed    0.02 sec
:info:test         Start  11: alternative_macros.cpp_xml
:info:test  11/135 Test  #11: alternative_macros.cpp_xml .................   Passed    0.02 sec
:info:test         Start  12: alternative_macros.cpp_junit
:info:test  12/135 Test  #12: alternative_macros.cpp_junit ...............   Passed    0.02 sec
:info:test         Start  13: assertion_macros.cpp
:info:test  13/135 Test  #13: assertion_macros.cpp .......................   Passed    0.02 sec
:info:test         Start  14: assertion_macros.cpp_xml
:info:test  14/135 Test  #14: assertion_macros.cpp_xml ...................   Passed    0.02 sec
:info:test         Start  15: assertion_macros.cpp_junit
:info:test  15/135 Test  #15: assertion_macros.cpp_junit .................   Passed    0.02 sec
:info:test         Start  16: stringification.cpp
:info:test  16/135 Test  #16: stringification.cpp ........................   Passed    0.02 sec
:info:test         Start  17: stringification.cpp_xml
:info:test  17/135 Test  #17: stringification.cpp_xml ....................   Passed    0.02 sec
:info:test         Start  18: stringification.cpp_junit
:info:test  18/135 Test  #18: stringification.cpp_junit ..................   Passed    0.02 sec
:info:test         Start  19: reporters_and_listeners.cpp
:info:test  19/135 Test  #19: reporters_and_listeners.cpp ................   Passed    0.02 sec
:info:test         Start  20: reporters_and_listeners.cpp_xml
:info:test  20/135 Test  #20: reporters_and_listeners.cpp_xml ............   Passed    0.02 sec
:info:test         Start  21: reporters_and_listeners.cpp_junit
:info:test  21/135 Test  #21: reporters_and_listeners.cpp_junit ..........   Passed    0.02 sec
:info:test         Start  22: subcases.cpp
:info:test  22/135 Test  #22: subcases.cpp ...............................   Passed    0.02 sec
:info:test         Start  23: subcases.cpp_xml
:info:test  23/135 Test  #23: subcases.cpp_xml ...........................   Passed    0.02 sec
:info:test         Start  24: subcases.cpp_junit
:info:test  24/135 Test  #24: subcases.cpp_junit .........................   Passed    0.02 sec
:info:test         Start  25: logging.cpp
:info:test  25/135 Test  #25: logging.cpp ................................   Passed    0.02 sec
:info:test         Start  26: logging.cpp_xml
:info:test  26/135 Test  #26: logging.cpp_xml ............................   Passed    0.02 sec
:info:test         Start  27: logging.cpp_junit
:info:test  27/135 Test  #27: logging.cpp_junit ..........................   Passed    0.02 sec
:info:test         Start  28: templated_test_cases.cpp
:info:test  28/135 Test  #28: templated_test_cases.cpp ...................   Passed    0.02 sec
:info:test         Start  29: templated_test_cases.cpp_xml
:info:test  29/135 Test  #29: templated_test_cases.cpp_xml ...............   Passed    0.02 sec
:info:test         Start  30: templated_test_cases.cpp_junit
:info:test  30/135 Test  #30: templated_test_cases.cpp_junit .............   Passed    0.02 sec
:info:test         Start  31: test_cases_and_suites.cpp
:info:test  31/135 Test  #31: test_cases_and_suites.cpp ..................   Passed    0.02 sec
:info:test         Start  32: test_cases_and_suites.cpp_xml
:info:test  32/135 Test  #32: test_cases_and_suites.cpp_xml ..............   Passed    0.02 sec
:info:test         Start  33: test_cases_and_suites.cpp_junit
:info:test  33/135 Test  #33: test_cases_and_suites.cpp_junit ............   Passed    0.02 sec
:info:test         Start  34: asserts_used_outside_of_tests.cpp
:info:test  34/135 Test  #34: asserts_used_outside_of_tests.cpp ..........   Passed    0.02 sec
:info:test         Start  35: asserts_used_outside_of_tests.cpp_xml
:info:test  35/135 Test  #35: asserts_used_outside_of_tests.cpp_xml ......   Passed    0.02 sec
:info:test         Start  36: asserts_used_outside_of_tests.cpp_junit
:info:test  36/135 Test  #36: asserts_used_outside_of_tests.cpp_junit ....   Passed    0.02 sec
:info:test         Start  37: enums.cpp
:info:test  37/135 Test  #37: enums.cpp ..................................   Passed    0.02 sec
:info:test         Start  38: enums.cpp_xml
:info:test  38/135 Test  #38: enums.cpp_xml ..............................   Passed    0.02 sec
:info:test         Start  39: enums.cpp_junit
:info:test  39/135 Test  #39: enums.cpp_junit ............................   Passed    0.02 sec
:info:test         Start  40: concurrency.cpp
:info:test  40/135 Test  #40: concurrency.cpp ............................   Passed    0.01 sec
:info:test         Start  41: concurrency.cpp_xml
:info:test  41/135 Test  #41: concurrency.cpp_xml ........................   Passed    0.01 sec
:info:test         Start  42: concurrency.cpp_junit
:info:test  42/135 Test  #42: concurrency.cpp_junit ......................   Passed    0.01 sec
:info:test         Start  43: namespace1.cpp
:info:test  43/135 Test  #43: namespace1.cpp .............................   Passed    0.01 sec
:info:test         Start  44: namespace1.cpp_xml
:info:test  44/135 Test  #44: namespace1.cpp_xml .........................   Passed    0.01 sec
:info:test         Start  45: namespace1.cpp_junit
:info:test  45/135 Test  #45: namespace1.cpp_junit .......................   Passed    0.01 sec
:info:test         Start  46: namespace2.cpp
:info:test  46/135 Test  #46: namespace2.cpp .............................   Passed    0.01 sec
:info:test         Start  47: namespace2.cpp_xml
:info:test  47/135 Test  #47: namespace2.cpp_xml .........................   Passed    0.01 sec
:info:test         Start  48: namespace2.cpp_junit
:info:test  48/135 Test  #48: namespace2.cpp_junit .......................   Passed    0.01 sec
:info:test         Start  49: namespace3.cpp
:info:test  49/135 Test  #49: namespace3.cpp .............................   Passed    0.02 sec
:info:test         Start  50: namespace3.cpp_xml
:info:test  50/135 Test  #50: namespace3.cpp_xml .........................   Passed    0.01 sec
:info:test         Start  51: namespace3.cpp_junit
:info:test  51/135 Test  #51: namespace3.cpp_junit .......................   Passed    0.01 sec
:info:test         Start  52: namespace4.cpp
:info:test  52/135 Test  #52: namespace4.cpp .............................   Passed    0.01 sec
:info:test         Start  53: namespace4.cpp_xml
:info:test  53/135 Test  #53: namespace4.cpp_xml .........................   Passed    0.01 sec
:info:test         Start  54: namespace4.cpp_junit
:info:test  54/135 Test  #54: namespace4.cpp_junit .......................   Passed    0.01 sec
:info:test         Start  55: namespace5.cpp
:info:test  55/135 Test  #55: namespace5.cpp .............................   Passed    0.01 sec
:info:test         Start  56: namespace5.cpp_xml
:info:test  56/135 Test  #56: namespace5.cpp_xml .........................   Passed    0.01 sec
:info:test         Start  57: namespace5.cpp_junit
:info:test  57/135 Test  #57: namespace5.cpp_junit .......................   Passed    0.01 sec
:info:test         Start  58: namespace6.cpp
:info:test  58/135 Test  #58: namespace6.cpp .............................   Passed    0.01 sec
:info:test         Start  59: namespace6.cpp_xml
:info:test  59/135 Test  #59: namespace6.cpp_xml .........................   Passed    0.01 sec
:info:test         Start  60: namespace6.cpp_junit
:info:test  60/135 Test  #60: namespace6.cpp_junit .......................   Passed    0.01 sec
:info:test         Start  61: namespace7.cpp
:info:test  61/135 Test  #61: namespace7.cpp .............................   Passed    0.01 sec
:info:test         Start  62: namespace7.cpp_xml
:info:test  62/135 Test  #62: namespace7.cpp_xml .........................   Passed    0.01 sec
:info:test         Start  63: namespace7.cpp_junit
:info:test  63/135 Test  #63: namespace7.cpp_junit .......................   Passed    0.01 sec
:info:test         Start  64: namespace8.cpp
:info:test  64/135 Test  #64: namespace8.cpp .............................   Passed    0.01 sec
:info:test         Start  65: namespace8.cpp_xml
:info:test  65/135 Test  #65: namespace8.cpp_xml .........................   Passed    0.01 sec
:info:test         Start  66: namespace8.cpp_junit
:info:test  66/135 Test  #66: namespace8.cpp_junit .......................   Passed    0.01 sec
:info:test         Start  67: namespace9.cpp
:info:test  67/135 Test  #67: namespace9.cpp .............................   Passed    0.01 sec
:info:test         Start  68: namespace9.cpp_xml
:info:test  68/135 Test  #68: namespace9.cpp_xml .........................   Passed    0.01 sec
:info:test         Start  69: namespace9.cpp_junit
:info:test  69/135 Test  #69: namespace9.cpp_junit .......................   Passed    0.01 sec
:info:test         Start  70: coverage_maxout.cpp
:info:test  70/135 Test  #70: coverage_maxout.cpp ........................   Passed    0.02 sec
:info:test         Start  71: coverage_maxout.cpp_xml
:info:test  71/135 Test  #71: coverage_maxout.cpp_xml ....................   Passed    0.02 sec
:info:test         Start  72: coverage_maxout.cpp_junit
:info:test  72/135 Test  #72: coverage_maxout.cpp_junit ..................   Passed    0.02 sec
:info:test         Start  73: version
:info:test  73/135 Test  #73: version ....................................   Passed    0.02 sec
:info:test         Start  74: version_xml
:info:test  74/135 Test  #74: version_xml ................................   Passed    0.02 sec
:info:test         Start  75: version_junit
:info:test  75/135 Test  #75: version_junit ..............................   Passed    0.02 sec
:info:test         Start  76: help
:info:test  76/135 Test  #76: help .......................................   Passed    0.02 sec
:info:test         Start  77: help_xml
:info:test  77/135 Test  #77: help_xml ...................................   Passed    0.02 sec
:info:test         Start  78: help_junit
:info:test  78/135 Test  #78: help_junit .................................   Passed    0.02 sec
:info:test         Start  79: outfile
:info:test  79/135 Test  #79: outfile ....................................   Passed    0.01 sec
:info:test         Start  80: outfile_xml
:info:test  80/135 Test  #80: outfile_xml ................................   Passed    0.01 sec
:info:test         Start  81: outfile_junit
:info:test  81/135 Test  #81: outfile_junit ..............................   Passed    0.01 sec
:info:test         Start  82: count
:info:test  82/135 Test  #82: count ......................................   Passed    0.02 sec
:info:test         Start  83: count_xml
:info:test  83/135 Test  #83: count_xml ..................................   Passed    0.02 sec
:info:test         Start  84: count_junit
:info:test  84/135 Test  #84: count_junit ................................   Passed    0.02 sec
:info:test         Start  85: list_test_cases
:info:test  85/135 Test  #85: list_test_cases ............................   Passed    0.02 sec
:info:test         Start  86: list_test_cases_xml
:info:test  86/135 Test  #86: list_test_cases_xml ........................   Passed    0.02 sec
:info:test         Start  87: list_test_cases_junit
:info:test  87/135 Test  #87: list_test_cases_junit ......................   Passed    0.02 sec
:info:test         Start  88: list_test_suites
:info:test  88/135 Test  #88: list_test_suites ...........................   Passed    0.02 sec
:info:test         Start  89: list_test_suites_xml
:info:test  89/135 Test  #89: list_test_suites_xml .......................   Passed    0.02 sec
:info:test         Start  90: list_test_suites_junit
:info:test  90/135 Test  #90: list_test_suites_junit .....................   Passed    0.02 sec
:info:test         Start  91: list_reporters
:info:test  91/135 Test  #91: list_reporters .............................   Passed    0.02 sec
:info:test         Start  92: list_reporters_xml
:info:test  92/135 Test  #92: list_reporters_xml .........................   Passed    0.02 sec
:info:test         Start  93: list_reporters_junit
:info:test  93/135 Test  #93: list_reporters_junit .......................   Passed    0.02 sec
:info:test         Start  94: all_binary
:info:test  94/135 Test  #94: all_binary .................................   Passed    0.02 sec
:info:test         Start  95: all_binary_xml
:info:test  95/135 Test  #95: all_binary_xml .............................   Passed    0.02 sec
:info:test         Start  96: all_binary_junit
:info:test  96/135 Test  #96: all_binary_junit ...........................   Passed    0.02 sec
:info:test         Start  97: abort_after
:info:test  97/135 Test  #97: abort_after ................................   Passed    0.02 sec
:info:test         Start  98: abort_after_xml
:info:test  98/135 Test  #98: abort_after_xml ............................   Passed    0.02 sec
:info:test         Start  99: abort_after_junit
:info:test  99/135 Test  #99: abort_after_junit ..........................   Passed    0.02 sec
:info:test         Start 100: first_last
:info:test 100/135 Test #100: first_last .................................   Passed    0.02 sec
:info:test         Start 101: first_last_xml
:info:test 101/135 Test #101: first_last_xml .............................   Passed    0.02 sec
:info:test         Start 102: first_last_junit
:info:test 102/135 Test #102: first_last_junit ...........................   Passed    0.02 sec
:info:test         Start 103: filter_1
:info:test 103/135 Test #103: filter_1 ...................................   Passed    0.02 sec
:info:test         Start 104: filter_1_xml
:info:test 104/135 Test #104: filter_1_xml ...............................   Passed    0.02 sec
:info:test         Start 105: filter_1_junit
:info:test 105/135 Test #105: filter_1_junit .............................   Passed    0.02 sec
:info:test         Start 106: filter_2
:info:test 106/135 Test #106: filter_2 ...................................   Passed    0.02 sec
:info:test         Start 107: filter_2_xml
:info:test 107/135 Test #107: filter_2_xml ...............................   Passed    0.02 sec
:info:test         Start 108: filter_2_junit
:info:test 108/135 Test #108: filter_2_junit .............................   Passed    0.02 sec
:info:test         Start 109: filter_3
:info:test 109/135 Test #109: filter_3 ...................................   Passed    0.02 sec
:info:test         Start 110: filter_3_xml
:info:test 110/135 Test #110: filter_3_xml ...............................   Passed    0.02 sec
:info:test         Start 111: filter_3_junit
:info:test 111/135 Test #111: filter_3_junit .............................   Passed    0.02 sec
:info:test         Start 112: order_1
:info:test 112/135 Test #112: order_1 ....................................   Passed    0.02 sec
:info:test         Start 113: order_1_xml
:info:test 113/135 Test #113: order_1_xml ................................   Passed    0.02 sec
:info:test         Start 114: order_1_junit
:info:test 114/135 Test #114: order_1_junit ..............................   Passed    0.02 sec
:info:test         Start 115: order_2
:info:test 115/135 Test #115: order_2 ....................................   Passed    0.02 sec
:info:test         Start 116: order_2_xml
:info:test 116/135 Test #116: order_2_xml ................................   Passed    0.02 sec
:info:test         Start 117: order_2_junit
:info:test 117/135 Test #117: order_2_junit ..............................   Passed    0.02 sec
:info:test         Start 118: order_3
:info:test 118/135 Test #118: order_3 ....................................   Passed    0.02 sec
:info:test         Start 119: order_3_xml
:info:test 119/135 Test #119: order_3_xml ................................   Passed    0.02 sec
:info:test         Start 120: order_3_junit
:info:test 120/135 Test #120: order_3_junit ..............................   Passed    0.02 sec
:info:test         Start 121: disabled
:info:test 121/135 Test #121: disabled ...................................   Passed    0.09 sec
:info:test         Start 122: disabled_xml
:info:test 122/135 Test #122: disabled_xml ...............................   Passed    0.02 sec
:info:test         Start 123: disabled_junit
:info:test 123/135 Test #123: disabled_junit .............................   Passed    0.02 sec
:info:test         Start 124: exe_with_static_libs
:info:test 124/135 Test #124: exe_with_static_libs .......................   Passed    0.11 sec
:info:test         Start 125: exe_with_static_libs_xml
:info:test 125/135 Test #125: exe_with_static_libs_xml ...................   Passed    0.02 sec
:info:test         Start 126: exe_with_static_libs_junit
:info:test 126/135 Test #126: exe_with_static_libs_junit .................   Passed    0.02 sec
:info:test         Start 127: executable_dll_and_plugin
:info:test 127/135 Test #127: executable_dll_and_plugin ..................***Failed    0.13 sec
:info:test         Start 128: executable_dll_and_plugin_xml
:info:test 128/135 Test #128: executable_dll_and_plugin_xml ..............***Failed    0.04 sec
:info:test         Start 129: executable_dll_and_plugin_junit
:info:test 129/135 Test #129: executable_dll_and_plugin_junit ............***Failed    0.04 sec
:info:test         Start 130: same_tests_multiple_configurations
:info:test 130/135 Test #130: same_tests_multiple_configurations .........***Failed    0.08 sec
:info:test         Start 131: same_tests_multiple_configurations_xml
:info:test 131/135 Test #131: same_tests_multiple_configurations_xml .....***Failed    0.01 sec
:info:test         Start 132: same_tests_multiple_configurations_junit
:info:test 132/135 Test #132: same_tests_multiple_configurations_junit ...***Failed    0.01 sec
:info:test         Start 133: playground
:info:test 133/135 Test #133: playground .................................   Passed    0.07 sec
:info:test         Start 134: playground_xml
:info:test 134/135 Test #134: playground_xml .............................   Passed    0.01 sec
:info:test         Start 135: playground_junit
:info:test 135/135 Test #135: playground_junit ...........................   Passed    0.01 sec
:info:test 96% tests passed, 6 tests failed out of 135
:info:test Total Test time (real) =   2.98 sec
:info:test The following tests FAILED:
:info:test 	127 - executable_dll_and_plugin (Failed)
:info:test 	128 - executable_dll_and_plugin_xml (Failed)
:info:test 	129 - executable_dll_and_plugin_junit (Failed)
:info:test 	130 - same_tests_multiple_configurations (Failed)
:info:test 	131 - same_tests_multiple_configurations_xml (Failed)
:info:test 	132 - same_tests_multiple_configurations_junit (Failed)
:info:test Errors while running CTest
:info:test Output from these tests are in: /opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_devel_doctest/doctest/work/doctest-2.4.6/Testing/Temporary/LastTest.log
:info:test Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
:info:test make: *** [test] Error 8
```

but we're in much better shape than we were =]

dankamongmen commented:
> okay, thanks for explaining; but then you might misunderstand the purpose of a PR - or at least the way we use it ;) Once you open a PR that signals that it's ready to review, in rare cases you could set it to `draft` but personally I am not a fan of this.... You're supposed to do the legwork locally and once you think everything works as intended you can open a PR.
> I will set this to a draft for now, once you've figured out things locally you can push to the PR and mark it as "ready for review".

absolutely, and i'm a bit embarrassed it happened -- i'd been up too long, and realized i needed this dep very late. sorry for this turd of a PR. things ought be rather more acceptable now, and i'm looking into these 6 failing unit tests.

dankamongmen commented on devel/doctest/Portfile:
```diff
@@ -0,0 +1,18 @@
+
+revision 0
+categories devel
+maintainers linux.com:nickblack
```
awesome, done, thanks!

dankamongmen commented:
Alrightie, I believe I've addressed all your feedback, which I will also go apply to my other outstanding PR (#11838). I'm going to try and elide the 6 failing unit tests, and at that point we ought be ready for PTAL.

dankamongmen commented:
hrmmm, it looks like eliding those unit tests would require creating a patch against the source (to remove them) -- there's no way to do it from the command line that i can see. would you prefer i make such a patch, or ought i just remove the `tests.run` line?

herbygillot commented:
> hrmmm, it looks like eliding those unit tests would require creating a patch against the source (to remove them) -- there's no way to do it from the command line that i can see. would you prefer i make such a patch, or ought i just remove the `tests.run` line?

You can make a patch, and include it in a `files/` directory alongside the Portfile.

Within the Portfile, you reference your patches using `patchfiles`.

Heres an [example of a port](https://github.com/macports/macports-ports/tree/master/net/newsboat) which uses patch files.

dankamongmen commented:
> Heres an [example of a port](https://github.com/macports/macports-ports/tree/master/net/newsboat) which uses patch files.

thanks much @herbygillot; i'll prep that up and do a final push. once this is in, i'll address your concerns regarding #11838.

dankamongmen commented:
alright i've got a working patch now; i'll integrate it into the PR

```c
diff -ur doctest-2.4.6/CMakeLists.txt doctest-2.4.6-patched/CMakeLists.txt
--- doctest-2.4.6/CMakeLists.txt	2021-03-22 08:05:04.000000000 -0400
+++ doctest-2.4.6-patched/CMakeLists.txt	2021-08-18 14:35:27.167498889 -0400
@@ -83,8 +83,6 @@
     # for code coverage we want exactly one binary to be produced and exercised
     if(NOT DEFINED ENV{CODE_COVERAGE})
         add_subdirectory(examples/exe_with_static_libs)
-        add_subdirectory(examples/executable_dll_and_plugin)
-        add_subdirectory(examples/combining_the_same_tests_built_differently_in_multiple_shared_objects)
         add_subdirectory(scripts/playground)
         add_subdirectory(examples/mpi)
     endif()
Only in doctest-2.4.6/examples: combining_the_same_tests_built_differently_in_multiple_shared_objects
Only in doctest-2.4.6/examples: executable_dll_and_plugin
```

herbygillot commented:
You may want to run the diff from within the root of the doctest dir, so that `doctest-2.4.6` is not embedded in the filenames within the patch.  Actually you can just edit the patchfile and change it to something like:
```
--- CMakeLists.txt	2021-03-22 08:05:04.000000000 -0400
+++ CMakeLists.txt	2021-08-18 14:35:27.167498889 -0400
@@ -83,8 +83,6 @@
...
```

That way it should apply in the Portfile, especially as the version of the port updates over time.

dankamongmen commented:
pushed the patch, recentered as described by @herbygillot. let's see if this flies...

reneeotten commented:
> pushed the patch, recentered as described by @herbygillot. let's see if this flies...

so this patch just removes the failing tests, correct? That does make the test-suite pass, but the fact these tests fail does that signal that anything is "wrong" with our installation or is it expected that they fail? 

Also, please do take care of aligning the fields in the Portfile correctly and follow the usual order.

reneeotten commented on devel/doctest/Portfile:
```diff
@@ -0,0 +1,24 @@
+PortGroup cmake 1.1
+PortGroup github 1.0
+github.setup        onqtam doctest 2.4.6
+version             2.4.6
```
remove the `version` field, it's redundant as it is already set from the line above by the github PG

reneeotten commented:
I also notice in the logs:
```
  -- Could NOT find MPI_CXX (missing: MPI_CXX_LIB_NAMES MPI_CXX_HEADER_DIR MPI_CXX_WORKS) 
  -- Could NOT find MPI (missing: MPI_CXX_FOUND CXX) 
```

to make sure that builds are reproducible you need to make sure that it builds the same irrespective of whether a user has a certain port installed. So if this is looking for MPI, it will likely use it when found - that's not what we want. Either you want to add a dependency on it and always use it, or add a configure argument that will avoid using it even when it is installed (in this case the latter would probably be most appropriate).

dankamongmen commented:
> to make sure that builds are reproducible you need to make sure that it builds the same irrespective of whether a user has a certain port installed. So if this is looking for MPI, it will likely use it when found - that's not what we want. Either you want to add a dependency on it and always use it, or add a configure argument that will avoid using it even when it is installed (in this case the latter would probably be most appropriate).

oh good catch. why on earth is doctest including OpenMP stuff?!? alright, i'll address these concerns when i can. investigating the test failures will probably take a bit of time (did any CI here exercise them originally, and have them not fail? i couldn't tell -- the failures i saw were local), so hopefully i'll post an update in a week or two (i don't have ready access to a macOS machine, so my chances to work on this are few and far between).

thanks for the solid feedback, @reneeotten !

dankamongmen commented:
interestingly, the tests only fail within the `brew test` environment; built outside, they work fine.

dankamongmen commented:
here's an example failure:

```
-- Found Git: /usr/bin/git (found version "2.30.1 (Apple Git-130)") 
diff --git a/opt/local/var/macports/build/_Users_dank_src_macports-ports_devel_doctest/doctest/work/doctest-2.4.6/examples/executable_dll_and_plugin/test_output/executable_dll_and_plugin_xml.txt b/opt/local/var/macports/build/_Users_dank_src_macports-ports_devel_doctest/doctest/work/build/examples/executable_dll_and_plugin/CMakeFiles/temp_test_output_executable_dll_and_plugin_xml.txt
index b8d3b4f..f386e3a 100644
--- a/opt/local/var/macports/build/_Users_dank_src_macports-ports_devel_doctest/doctest/work/doctest-2.4.6/examples/executable_dll_and_plugin/test_output/executable_dll_and_plugin_xml.txt
+++ b/opt/local/var/macports/build/_Users_dank_src_macports-ports_devel_doctest/doctest/work/build/examples/executable_dll_and_plugin/CMakeFiles/temp_test_output_executable_dll_and_plugin_xml.txt
@@ -1,61 +1,3 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<doctest binary="executable_dll_and_plugin">
-  <Options order_by="file" rand_seed="0" first="0" last="4294967295" abort_after="0" subcase_filter_levels="2147483647" case_sensitive="false" no_throw="false" no_skip="false"/>
-  <TestSuite>
-    <TestCase name="dll" filename="dll.cpp" line="0">
-I am a test from the dll!
-      <OverallResultsAsserts successes="0" failures="0"/>
-    </TestCase>
-    <TestCase name="implementation" filename="implementation.cpp" line="0">
-I am a test from the implementation!
-      <OverallResultsAsserts successes="0" failures="0"/>
-    </TestCase>
-    <TestCase name="implementation_2" filename="implementation_2.cpp" line="0">
-I am a test from the implementation_2!
-      <OverallResultsAsserts successes="0" failures="0"/>
-    </TestCase>
-    <TestCase name="executable" filename="main.cpp" line="0">
-I am a test from the executable!
-      <Exception crash="false">
-        char: 97
-      </Exception>
-      <OverallResultsAsserts successes="0" failures="0"/>
-    </TestCase>
-  </TestSuite>
-  <TestSuite name="some test suite">
-    <TestCase name="test case in a plugin" filename="plugin.cpp" line="0">
-      <SubCase name="some subcase" filename="plugin.cpp" line="0">
-        <Message type="WARNING" filename="plugin.cpp" line="0">
-          <Text>
-            triggering the INFO above to be printed
-          </Text>
-          <Info>
-            some info
-          </Info>
-        </Message>
-        <Expression success="false" type="CHECK" filename="plugin.cpp" line="0">
-          <Original>
-            1 == 2
-          </Original>
-          <Expanded>
-            1 == 2
-          </Expanded>
-          <Info>
-            some info
-          </Info>
-        </Expression>
-        <Message type="FATAL ERROR" filename="plugin.cpp" line="0">
-          <Text>
-            certain death!
-          </Text>
-          <Info>
-            some info
-          </Info>
-        </Message>
-      </SubCase>
-      <OverallResultsAsserts successes="0" failures="2"/>
-    </TestCase>
-  </TestSuite>
-  <OverallResultsAsserts successes="0" failures="2"/>
-  <OverallResultsTestCases successes="3" failures="2" skipped="0"/>
-</doctest>
+dyld: Library not loaded: /opt/local/lib/libdll.dylib
+  Referenced from: /opt/local/var/macports/build/_Users_dank_src_macports-ports_devel_doctest/doctest/work/build/examples/executable_dll_and_plugin/executable_dll_and_plugin
+  Reason: image not found

CMake Error at /opt/local/var/macports/build/_Users_dank_src_macports-ports_devel_doctest/doctest/work/doctest-2.4.6/scripts/cmake/exec_test.cmake:69 (message):
  Running
  '/opt/local/var/macports/build/_Users_dank_src_macports-ports_devel_doctest/doctest/work/build/examples/executable_dll_and_plugin/executable_dll_and_plugin
  --no-version --reporters=xml --dt-no-path-filenames=1
  --dt-no-line-numbers=1 --dt-no-exitcode=1 --dt-gnu-file-line=0
  --dt-no-time-in-output=1' ended with code 'Output is different from
  reference file!'


<end of output>
Test time =   0.03 sec
```

dankamongmen commented:
ok, when i run the offending test from within the `work/build/` subdirectory, i reproduce this output:

```
[cobra-la](134) $ pwd
/Users/dank/src/macports-ports/devel/doctest/work/build
[cobra-la](0) $  examples/executable_dll_and_plugin/executable_dll_and_plugin 
dyld: Library not loaded: /opt/local/lib/libdll.dylib
  Referenced from: /opt/local/var/macports/build/_Users_dank_src_macports-ports_devel_doctest/doctest/work/build/examples/executable_dll_and_plugin/executable_dll_and_plugin
  Reason: image not found
Abort trap: 6
[cobra-la](134) $ 
```

dankamongmen commented:
working:

```
[cobra-la](127) $ otool -L build/examples/executable_dll_and_plugin/executable_dll_and_plugin 
build/examples/executable_dll_and_plugin/executable_dll_and_plugin:
	@rpath/libdll.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libimplementation.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)
[cobra-la](0) $
```

not working (macports):

```
[cobra-la](0) $ cd -
/Users/dank/src/macports-ports/devel/doctest/work/build
[cobra-la](0) $ otool -L examples/executable_dll_and_plugin/executable_dll_and_plugin 
examples/executable_dll_and_plugin/executable_dll_and_plugin:
	/opt/local/lib/libdll.dylib (compatibility version 0.0.0, current version 0.0.0)
	/opt/local/lib/libimplementation.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)
[cobra-la](0) $ 
```

dankamongmen commented:
so it seems `/opt/local/lib/libdll.dylib` ought be `@rpath/libdll.dylib`. i'm not very intimately familiar with macOS -- does anyone see anything obvious i ought be doing here?

dankamongmen commented:
> to make sure that builds are reproducible you need to make sure that it builds the same irrespective of whether a user has a certain port installed. So if this is looking for MPI, it will likely use it when found - that's not what we want. Either you want to add a dependency on it and always use it, or add a configure argument that will avoid using it even when it is installed (in this case the latter would probably be most appropriate).

there doesn't seem to be any such configuration option at the moment, so i'll have to submit a patch upstream.

this is kinda spiraling into more work than i'd really been looking to do, with a few moving parts. i'll just withdraw this port. thanks for all your help!

reneeotten commented:
> this is kinda spiraling into more work than i'd really been looking to do, with a few moving parts. i'll just withdraw this port. thanks for all your help!

well... I also spent enough time on this now to help you out and it would be waste not to get this finished. Let me take a closer look at it myself later today. You said above it's needed for the `notcurses` port, but I don't see a dependency on it in that PR - what was the reason for this again?

mojca commented:
@dankamongmen: I would be in favour of reopening the pull request as well, so that someone else can fix the remaining issues, even if you don't plan to finish the work.

Doctest is really really cool, I would support keeping it inside macports even if we don't directly need it for another port.

I just wanted to mention that we have other ports failing some tests in the test suite, so it's not the end of the world if not all the tests pass.

I'm usually working with a single header file when I use doctest, but I admit that I had quite some issues getting the discovery of the library work correctly via the CMake mechanism (I actually never got it to work; I wrote my own CMake support for my own purposes, from scratch).

According to my experience the author of the code is super responsive to fix any issues, at least in the C++ part of the code.

reneeotten commented:
@dankamongmen I just pushed this to master with a few minor changes. Most of the tests pass and for the ones that fail it's just looking in the wrong place for the `.dylib` files; I am not sure how to fix that... Running the test-suite isn't necessarily a prerequisite, so let's assume that it all works correctly. 

Turns out I was wrong about the MPI stuff... looking a bit closer this port installs only a header (+ cmake files), so it's not linking to anything. It is just checking whether mpi is used so that it can include the correct stuff... So I think this is all good.

dankamongmen commented:
oh, certainly feel absolutely free to use what i had here as a starting base! i maintain doctest for a few other distributions (Alpine Linux, FreeBSD  (and thus DPorts), etc.) so i figured i could just kinda sweep MacPorts up into things. my lack of ready access to an actual macintosh was making this frustrating.

i applaud your high standards, and appreciate your kind feedback and instruction. i'm delighted to see doctest merged.

i will happily submit version bumps to MacPorts while doing my normal rounds at doctest release time.
